---
title: "Integration of DMPs and PCHiC"
author: "Andrea Nieto-Aliseda Sutton"
date: "06/04/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(GenomicInteractions)
library(JavierreLabR)
library(org.Hs.eg.db)
library(clusterProfiler)
library(BSgenome.Hsapiens.UCSC.hg19)
```


## STEP 1.3
### Integration  

We have <span style="color: red;">225,785 DMPs</span> from out differential methylation analysis. These must be integrated with our enhancer regions in order to obtain a list of DMPs that fall within enhancers.  

```{r, echo=FALSE}
enhancers <- read.table("/media/andrea/SAMSUNG/IJC/Regulatory_Builds/general/rmarkdown/enhancer_regions.txt", stringsAsFactors = F)
colnames(enhancers) <- c("ID","chr","start","end")
enhancers <- arrange(enhancers, chr, start)
e <- makeGRangesFromDataFrame(enhancers, keep.extra.columns = T)


dmps <- as_tibble(read.table("/media/andrea/SAMSUNG/IJC/Aim2/methylation/FUNCTIONALNORMALISATION/DMPs_with_DMstatus.txt", header = T, stringsAsFactors = F)) #reading dmps
dmps <- dmps[,c(4,30,1,2)]
dmps <- cbind(dmps,dmps[,4])
colnames(dmps) <- c("name","dm_status","chr","start","end")

d <- makeGRangesFromDataFrame(dmps, seqnames.field = "chr", start.field = "start", end.field = "end", keep.extra.columns = T) #dmps as GRanges
  
#overlap to find enhancers
overlape <- findOverlaps(d, e)
overlape <- e[subjectHits(overlape),]
uniqueoverlape <- unique(overlape)  #5234 unique enhancers

#overlap to find dmps
overlapd<- findOverlaps(e, d)
overlapd <- d[subjectHits(overlapd),] #5837 dmps fall in known enhancers
uniqueoverlapd <- unique(overlapd) #obviously, one DMP can only fall in one enhancer
uniqueoverlapd

```

Above we see an excerpt of the DMPs falling in enhancer regions. This is a total of 5837.  
The number of enhancer regions involved is 5,234. This means that some enhancers have more than one DMP.  

```{r, echo=FALSE}

enhancer_dmps <- as.data.frame(overlapd@elementMetadata)
keep <- dmps$name %in% enhancer_dmps$name
enhancer_dmps <- dmps[keep,]

enhancer_dmps <- arrange(enhancer_dmps, chr, start)
enhancer_dmps <- enhancer_dmps[,c(1,3,4,5,2)]

```
  
  
## STEP 2.1  
### Data  

This next step involves the integration of enhancer DMPs with PCHi-C data (be that of Naive B, HSC, pre-B or pro-B). For now the only PCHi-C we have available is that of Naive B, which we have to load into the session using the novel JavierreLabR package :)  

```{r , include=FALSE}

pchic <- JavierreLabR::load_interactions("/media/andrea/SAMSUNG/IJC/NB/NB_Merged_Step2.txt_seqmonk.txt")
pchic <- JavierreLabR::annotate_interactions(interactions = pchic, annotation = "/media/andrea/SAMSUNG/IJC/superquery/digest_and_probes_homo_sapiens_hg19_updated_16_02_2020.txt")

dat <- pchic@elementMetadata
pchic <- pchic[dat$int == "P_OE"|dat$int == "P_P",]
dat <- dat[dat$int == "P_OE"|dat$int == "P_P",]
pchic <- pchic[!(dat$gene_I == "non-annotated"),]
dat <- dat[!(dat$gene_I == "non-annotated"),]
pchic <- pchic[dat$gene_II != "non-annotated",]
dat <- dat[!(dat$gene_II == "non-annotated"),]

```
  
  
A total of 192,104 interactions are annotated from the Naive-B cell PCHi-C data.  
However, for now, we will exclude from the analysis any interaction involving fragments enominated as UCEs or non-annotated. This leaves us with a <span style="color: red;">total of 189284 interactions to work with</span>

  
## STEP 2.2 
### Integration  

The PCHi-C interaction data and the enhancer DMPs are overlapped at this point, also using JavierreLabR package.
```{r , include=FALSE}

aux <- enhancer_dmps[,c(2,3,4)]
#write.table(aux, "/media/andrea/SAMSUNG/IJC/Aim3/WORKFLOW-31-03/auxtable.txt", col.names = F, row.names = F, quote = F, sep = "\t")
integration <- JavierreLabR::integrate_regions(interactions = pchic, regions = "/media/andrea/SAMSUNG/IJC/Aim3/WORKFLOW-31-03/auxtable.txt")
```
<span style="color: red;">Integration of Naive-B interactions with enhancer DMPs results in 4284 interactions.</span>  
  
  
  
## STEP 3  
### Exploration  

**What types of interactions to we have in this subset?**
```{r , echo=FALSE}

table(integration@elementMetadata$int)

```
  
Knowing that the DMPs fall in enhancers, we can assume that those interactions with peaks in P (captured regions) dont actually fall within promoter regions.  

The only types of interactions that will not be taken into account are the P_OE_I.  

We have two subgroup of interactions:  
1. OE_P_I + P_OE_II (970 interactions)  
2. P_P_I + P_P_II (617 interactions)  
  
  
  
## STEP 4
### Listing possible transcripts/genes being dysregulated

> SUGROUP 1  

```{r , echo=FALSE}

subset <- integration[integration@elementMetadata$int == "OE_P_I" | integration@elementMetadata$int == "P_OE_II",]

d <- subset@elementMetadata

t <- c(unlist(strsplit(d$gene_I[d$overlap_II == TRUE], ",")))

#write(t, file = "/media/andrea/SAMSUNG/IJC/Aim3/WORKFLOW-31-03/transcripts_subgroup1.txt")

superquery <- read.table("/media/andrea/SAMSUNG/IJC/superquery/realfinal_superquery.txt", header = T, stringsAsFactors = F)
g <- unique(superquery$ensembl_gene_id[superquery$ensembl_transcript_id %in% t])

#write(g, file = "/media/andrea/SAMSUNG/IJC/Aim3/WORKFLOW-31-03/genes_subgroup1.txt")

```
Number of transcripts possibly being dysregulated by DMPs at long-range enhancers: 4632.  
This translates to <span style="color: red;">1107 genes</span>  
  
  
> SUGROUP 2  

```{r , echo=FALSE}

subset <- integration[integration@elementMetadata$int == "P_P_I" | integration@elementMetadata$int == "P_PP_II",]

d <- subset@elementMetadata

t <- c(unlist(strsplit(d$gene_I[d$overlap_II == TRUE], ",")))

#write(t, file = "/media/andrea/SAMSUNG/IJC/Aim3/WORKFLOW-31-03/transcripts_subgroup2.txt")

superquery <- read.table("/media/andrea/SAMSUNG/IJC/superquery/realfinal_superquery.txt", header = T, stringsAsFactors = F)
g <- unique(superquery$ensembl_gene_id[superquery$ensembl_transcript_id %in% t])

#write(g, file = "/media/andrea/SAMSUNG/IJC/Aim3/WORKFLOW-31-03/genes_subgroup2.txt")

```
Number of transcripts possibly being dysregulated by DMPs at long-range enhancers: 2015    
This translates to <span style="color: red;">500 genes</span>  
  
  
Subgroup   | Interactions    | N^o^ transcripts | N^o^ genes
-----------|-----------------|------------------|-------------
1          | OE_P_I, P_OE_II | 4632             | 1107
2          | P_P_I, P_P_II   | 2015             | 500
both       | all above       | 6647             | 1607  
  
  